### PR TITLE
Fetch strategy packages into a durable absolute root, never CWD-relative

### PR DESCRIPTION
**Pull request by danielmbirochi**
## Problem

`deploy.py`/`close.py` fetched remote strategy packages into a **CWD-relative** `strategies/` dir. When the agent shell's CWD was inside a managed skill dir (a common place to run the scripts from), the package — and its `.deploy-state.json` — landed inside a directory the runtime's skills-manager swap-replaces on every SKILL.md version bump. Those deployments were silently destroyed and their scanners died at the next gateway restart (`scanner mount failed — entrypoint not found`, the 2026-07-30 fleet incident). Whether a user's deploy survived was a pure CWD lottery — and the authoring flow had the same lottery, since the author skill scaffolded at CWD-relative `strategies/<id>` too.

Ticket: https://app.notion.com/p/3ae1feaac80681478ddde5d6a307648a

## Fix

- New `_pkg.strategies_root()` — the single durable, CWD-independent root: `SENPI_STRATEGIES_DIR` env override → `<OPENCLAW_WORKSPACE_DIR>/strategies` (the workspace is relocatable) → `/data/workspace/strategies` (agent-host default) → legacy CWD-relative `strategies/` for dev hosts, now with a **loud stderr warning** (it reintroduces the exact CWD-dependence this exists to remove; a relative `SENPI_STRATEGIES_DIR` warns too).
- `deploy.py ensure_pkg` and `close.py` fetch into that root and load by explicit absolute path — no CWD dependence.
- **Bare-id resolution prefers the durable root over CWD** (arg-as-is → durable → CWD-relative legacy fallback): the durable copy holds the deploy state, and letting a pristine checkout or stale skill-dir copy shadow it can double-fund a wallet or register a runtime against a dead one. Pre-durable-root deploys that only exist CWD-relative are still found via the fallback.
- **Author flow fixed** (the remaining wipe vector): `senpi-strategy-author/SKILL.md` + `references/creating-a-strategy.md` now scaffold and validate at `/data/workspace/strategies/<id>` with an explicit never-inside-a-skill-dir rule. Author bumped to **2.11.0**.
- Ops `SKILL.md` + `references/lifecycle.md` document the root and the resolution order; resolver error messages name every location actually tried. Ops bumped to **2.10.0**.
- Tests (111 passing): env/workspace tiers, loud fallback, durable-from-any-CWD, durable-beats-CWD shadowing, legacy CWD fallback, and the incident case itself (fetch with CWD = skill dir writes to the durable root, nothing lands in the skill dir).

## Review

Adversarially reviewed by two independent passes; addressed in `b2fbb22a` (workspace tier, loud fallbacks, durable-first precedence, author-skill guidance, error messages). Known residuals, deliberate: an on-disk package remains authoritative (`--ref` never overwrites a resolvable local copy — pre-existing design); `close.py`'s invalid-local → remote-fetch fallback is pre-existing; already-deployed packages inside skill dirs are not migrated by this PR (the restore initiative owns relocation — until then they're protected by the runtime preserve-list).

## ⚠️ Merge ordering

**Do not merge until Senpi-ai/senpi-trading-runtime#273 is released and fleet-converged** (verify via ClickStack `service.version` distribution). Merging this bumps ops to 2.10.0 and author to 2.11.0, triggering fleet-wide reinstalls of both skill dirs — the exact operation that wipes user data on boxes still running a runtime without the preserve-list fix.

---

> [!NOTE]
> **High Risk**
> Changes where packages resolve and which copy wins when duplicates exist — wrong resolution can double-fund or render the wrong runtime onto live wallets; fleet skill bumps can still wipe pre-migration packages until runtime #273 is deployed.
> 
> **Overview**
> **Strategy packages and deploy state must live outside managed skill dirs** — CWD-relative `strategies/` caused fetches and authored packages to land inside wipeable skill trees on the next version bump.
> 
> Ops adds **`_pkg.strategies_root()`** (`SENPI_STRATEGIES_DIR` → `OPENCLAW_WORKSPACE_DIR/strategies` → `/data/workspace/strategies` → legacy CWD with stderr warnings). **`deploy.py`**, **`close.py`**, and **`validate_universe.py --all`** fetch and enumerate from that root. **`resolve_pkg_dir`** prefers the durable copy, uses **`.deploy-state.json`** when durable and CWD copies conflict, and keeps a CWD legacy fallback. **`_fetch._out_path`** blocks path traversal on remote tree entries.
> 
> **Author skill 2.11.0** scaffolds and validates at `/data/workspace/strategies/<id>` with an explicit never-inside-skill-dir rule; **ops 2.10.0** docs mirror the resolution order. Tests lock root tiers, shadowing, fetch-from-skill-CWD, and traversal refusal.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 0ec7e88ad17562d3a82978eb0b11ff630d004354. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Review by dmcr — APPROVED** [review]

**Comment by 0xsarvesh** [comment]
Reviewed the full diff and ran the suite plus a set of edge-case probes. The durable-root approach is right and the tests in the PR are good. **Two findings I'd want fixed before this merges** — one silently reopens the exact hole this PR closes, and one can cost money during the restoration.

Suite on `fix/durable-strategy-dest-root`: **105 passed, 0 failed**, matching the PR body.

I then wrote 6 probes for cases the PR's tests don't cover. Three passed (empty/whitespace env hygiene, absolute-arg join through the durable tier, trailing-slash arg). Three failed.

## 1. The durable root falls back to CWD when the workspace dir doesn't exist yet — `_pkg.py:172`

```python
if workspace_env and Path(workspace_env).is_dir():
```

`OPENCLAW_WORKSPACE_DIR=/data/workspace` is set in production (`senpi-hyperclaw-railway-template/railway.toml:15`), but this tier is gated on the directory *already existing*. On a fresh volume — or any host where openclaw creates the workspace lazily — the env tier fails the `is_dir()` check, the hardcoded `/data/workspace` tier fails it too, and `strategies_root()` returns `Path("strategies")`: CWD-relative. That is precisely the incident, reintroduced on the boxes least likely to be watched, and the loud warning fires into a shell nobody is reading.

Probe:

```python
monkeypatch.delenv("SENPI_STRATEGIES_DIR", raising=False)
ws = tmp_path / "not-created-yet"          # env SET, dir absent
monkeypatch.setenv("OPENCLAW_WORKSPACE_DIR", str(ws))
assert _pkg.strategies_root() == ws / "strategies"   # FAILS -> returns Path("strategies")
```

Setting the variable is an explicit declaration of intent; it should be honored regardless of whether the directory has materialized yet — `_fetch.py:76` already does `mkdir(parents=True, exist_ok=True)` on write. Suggest dropping the `is_dir()` guard on the env tier and keeping it only on the hardcoded `/data/workspace` tier, where it's legitimate host detection.

## 2. A pristine fetch permanently shadows a legacy deploy's state — can double-fund a wallet

This is the one I'd most want addressed before the restoration runs.

1. An impacted user's real deploy sits at the legacy `<skilldir>/strategies/spider`, carrying `.deploy-state.json` with a **live wallet**.
2. The agent runs any bare-id ops command from a CWD that isn't that skill dir. Resolution misses all three tiers, so `ensure_pkg` fetches a **pristine** copy from the catalog into the durable root — no `.deploy-state.json`.
3. Durable-first precedence now makes that pristine copy win from **every** CWD, permanently.
4. A later `deploy.py create spider --budget N` sees no deploy state, and funds a **second** wallet while the first is still live.

Probe `test_probe_pristine_durable_shadows_legacy_deploy_state` fails on exactly this: after the fetch, resolving the bare id from the legacy CWD returns the durable copy with no `.deploy-state.json`.

Worth being precise about what's new here: the pristine-fetch-shadows-real-deploy hazard is pre-existing. What this PR changes is its blast radius — before, the pristine copy landed CWD-relative, so the damage was transient and scoped to one directory; with durable-first it becomes permanent and global. And it lands specifically on the 24+ users being restored, who are the ones with packages in legacy locations.

Two mitigations, ideally both:

- **In code:** when more than one copy resolves, prefer the one carrying `.deploy-state.json`. It's the field that actually distinguishes "the deployed package" from "a checkout of the same id", which is what the precedence rule is really trying to express.
- **Operationally:** migrate legacy packages into the durable root *before* any bare-id command runs during restoration, so step 2 above can't happen.

## 3. `validate_universe.py:104` still globs CWD-relative

```python
pkgs = sorted(d for d in glob.glob("strategies/*") if os.path.isfile(os.path.join(d, "strategy.yaml")))
```

Missed site, same bug class. Read-only and it fails loudly (`ap.error`) rather than silently passing, so no data loss — but after this PR `--all` scans the wrong root. One line: glob `_pkg.strategies_root()`.

## 4. Minor: `_fetch.py:75` destination traversal

`out = Path(dest_root) / path[len("strategies/"):]` builds the write path from remote tree entries. The arithmetic does escape `dest_root` given a `..` segment, but git won't emit `..` in tree paths, so this is only reachable via a hostile `SENPI_SKILLS_REPO`. Defense-in-depth rather than a live vuln — a `resolve()` + `is_relative_to(dest_root)` assert would close it cheaply.

## Merge ordering

Agreed on the ordering, and the warning in the PR body is the right call. One thing I'd want nailed down before this merges: Senpi-ai/senpi-trading-runtime#273 ships via the runtime's npm auto-update, and the skills-manager is itself the subsystem that performs updates — so boxes with a pinned or broken auto-updater never converge, and merging this bumps ops/author and wipes exactly those. What convergence threshold gates the merge, and what happens to the tail that never gets there?

File: senpi-strategy-ops/scripts/_pkg.py [hunk 1]
```diff
@@ -148,12 +150,53 @@ def any_needs_model(self):
...
+            print(f"⚠ SENPI_STRATEGIES_DIR={env!r} is a RELATIVE path — packages there may not "
+                  f"survive skill updates; use an absolute path.", file=sys.stderr)
+        return root
+    # A SET workspace env is honored even before the dir exists (fresh volume, gateway not yet
+    # booted — the gateway mkdir -p's it at boot, and fetch mkdir -p's on write): gating on
+    # is_dir() would fall through to CWD-relative, the exact incident behavior. Existence-gating
+    # is only legitimate on the hardcoded tier below, where it's host detection.
+    workspace_env = os.environ.get("OPENCLAW_WORKSPACE_DIR", "").strip()
+    if workspace_env:
+        return Path(workspace_env) / "strategies"
```
[the remark this answers is not in the document]
Comment (danielmbirochi) [reply senpi-strategy-ops/scripts/_pkg.py:180]: Fixed in dc34d236: a relative `OPENCLAW_WORKSPACE_DIR` now warns exactly like a relative `SENPI_STRATEGIES_DIR` — honor-but-warn on both env tiers (test: `test_strategies_root_relative_workspace_env_warns`). Honoring rather than skipping is deliberate and matches the existing `SENPI_STRATEGIES_DIR` contract: an explicitly set env var wins, the warning is the guardrail. In production the platform sets it absolute (`railway.toml` → `/data/workspace`).

File: senpi-strategy-ops/scripts/_pkg.py [hunk 2]
```diff
@@ -148,12 +150,53 @@ def any_needs_model(self):
...
+    <strategies_root()>/<arg> (the durable root where remote fetches land — authoritative for bare
+    ids, since it holds the deploy state), then strategies/<arg> (CWD-relative, legacy fallback for
+    pre-durable-root deploys). Durable-before-CWD matters: a pristine repo checkout or a stale
+    skill-dir copy must not shadow the deployed package and its .deploy-state.json."""
     p = Path(arg)
     if (p / "strategy.yaml").is_file():
         return p
+    durable = strategies_root() / arg
+    if (durable / "strategy.yaml").is_file():
+        return durable
```
[the remark this answers is not in the document]
Comment (danielmbirochi) [reply senpi-strategy-ops/scripts/_pkg.py]: Fixed in 4c7d1629: when both the durable and CWD-relative copies exist, the copy carrying `.deploy-state.json` now wins (ties stay durable; two copies both carrying state warns loudly). Tests: `test_resolve_pkg_dir_deploy_state_beats_pristine_durable`, `test_resolve_pkg_dir_both_have_state_durable_wins_loudly`.

One correction to the failure narrative: `create` does not close funded wallets off local state — it reconciles against the backend by `skillName` first (`deploy.py` `cmd_create`), refusing outright when a running runtime exists and only closing (funds recovered) runtime-less open strategies, by design. The real exposure from the shadowing was `runtime`: it self-heals the live wallet from the backend and would have rendered the pristine copy's catalog-default `runtime.yaml` onto it.

**Comment by danielmbirochi** [comment]
All four findings addressed on the branch (suite 105 → 111, all passing):

**1. Env tier gated on `is_dir()` — fixed in 26e3bd55.** A set `OPENCLAW_WORKSPACE_DIR` is now honored unconditionally; existence-gating stays only on the hardcoded `/data/workspace` tier where it's host detection. Your probe passes as written. One scoping note: the gateway `mkdir -p`s the workspace at boot (`senpi-agent/src/gateway.js`), so on a healthy prod box the old gate always passed — the realistic exposure was scripts run outside/before the gateway (recovery shells, dev boxes). The contract argument stands regardless. Follow-up dc34d236 closes the gap this opened for a *relative* env value (warns, matching the `SENPI_STRATEGIES_DIR` tier).

**2. Pristine durable copy shadowing a legacy deploy — fixed in 4c7d1629.** When both the durable and CWD copies exist, the copy carrying `.deploy-state.json` wins; ties stay durable, and two copies both carrying state warns loudly. Explicit paths remain authoritative.

Two precision points on the blast radius. First, the double-fund terminal step was already guarded independently of local state: `cmd_create` reconciles against the backend by `skillName` before funding — it refuses when a running runtime exists on an open wallet, and closes (recovering funds) runtime-less ones. A true double-fund needs the backend lookup to miss the legacy strategy. The live hazard the shadowing actually amplified is `runtime`: it self-heals the correct live wallet from the backend, then would have rendered the pristine copy's catalog-default `runtime.yaml` onto it — replacing tuned parameters on live money. Second, the honest limit of the code fix: it helps only where the legacy copy is visible at resolution time (its own CWD). From any other directory no resolution rule can find a CWD-relative package, so your operational mitigation is the complete fix — the restoration plan now carries the ordering constraint (migrate legacy packages into the durable root before any bare-id command runs on a box).

**3. `validate_universe --all` CWD glob — fixed in 12cc9c2b.** Now enumerates `_pkg.strategies_root()` (new `all_packages()` helper, tested); dev hosts keep legacy behavior via the root's own CWD fallback.

**4. Fetch dest traversal — added in e00485d3.** `_out_path()` refuses tree entries escaping the dest root.

**Merge ordering.** The gate we're using: not whole-fleet convergence, but **zero boxes that both (a) still run a pre-preserve-list runtime and (b) have strategy packages inside skill dirs** — a box with nothing in a skill dir has nothing to lose from the bump. The never-converging tail (pinned/broken auto-updaters, overlapping heavily with the chronic pre-Jul-22 cohort) gets handled per-box rather than holding the merge: migrate their packages into the durable root manually, or roll the image directly. The `service.version` distribution measurement is running now; merge waits on that gate, and the restoration executes after the merge so operators run the fixed scripts.

File: senpi-strategy-ops/scripts/_fetch.py [hunk 3]
```diff
@@ -42,6 +45,16 @@ def _get(host, path, accept, timeout):
...
 
 
+def _out_path(dest_root, tree_path):
+    """Local dest for a remote tree entry (strategies/<id>/...), refusing any path that escapes
+    dest_root. Defense-in-depth: git won't emit `..` in tree paths, but the repo/ref this fetches
+    from are env-overridable (SENPI_SKILLS_REPO/_REF)."""
+    out = Path(dest_root) / tree_path[len("strategies/"):]
+    if not out.resolve().is_relative_to(Path(dest_root).resolve()):
+        raise FetchError(f"remote tree entry {tree_path!r} escapes the dest root — refusing")
+    return out
```
[the remark this answers is not in the document]
Comment (danielmbirochi) [reply senpi-strategy-ops/scripts/_fetch.py:57]: Fixed in 0ec7e88a — switched to the 3.8-safe `relative_to` + `ValueError` idiom. Verified both halves: production boxes are unaffected either way (the senpi-agent Railway image's runtime stage is `node:22-bookworm` → Debian 12 → Python 3.11), but README does document agent hosts as Python 3.8+ and this was the repo's only 3.9+ usage, so the documented floor stays honest for self-hosted hosts.

File: senpi-strategy-ops/scripts/_pkg.py [hunk 4]
```diff
@@ -148,14 +150,73 @@ def any_needs_model(self):
...
+    remote fetches land), then strategies/<arg> (CWD-relative, legacy fallback for pre-durable-root
+    deploys). When BOTH the durable and CWD copies exist, the one carrying .deploy-state.json wins —
+    that file is what distinguishes THE DEPLOYED PACKAGE from a checkout of the same id. Either
+    direction of shadowing is money-adjacent: a pristine copy over the deployed one makes `runtime`
+    render catalog-default parameters onto a live recovered wallet; ties go durable (and two copies
+    BOTH carrying state is a real ambiguity — warned loudly, the backend reconcile owns wallet truth)."""
     p = Path(arg)
     if (p / "strategy.yaml").is_file():
         return p
+    durable = strategies_root() / arg
```
[the remark this answers is not in the document]
Comment (danielmbirochi) [reply senpi-strategy-ops/scripts/_pkg.py:203]: Confirmed and fixed in ad81b43a — this was the most severe of the review findings: unlike the shadowing case there's only one copy, so the fetch grafts pristine files onto the deployed package **in place** and no resolution preference can recover it.

Fix restores the invariant that resolution always tries exactly the location a fetch would write to: both bare-id tiers now key on `Path(arg).name` (the strategy id, which is also `ensure_pkg`'s fetch key), so `strategies/<id>`, `<id>`, and any path spelling of the same id resolve identically; error messages name the real locations tried. Defense-in-depth for the remaining edge (dir carrying `.deploy-state.json` but no loadable `strategy.yaml`, e.g. a partially wiped package): `ensure_pkg` now refuses to fetch into it rather than overwriting. Tests: `test_resolve_pkg_dir_path_form_finds_durable_package`, `test_ensure_pkg_never_fetches_over_deploy_state` (suite at 113).

[the remark this answers is not in the document]
Comment (danielmbirochi) [reply senpi-strategy-ops/scripts/_pkg.py:203]: Correction to the trail: this fix (ad81b43a) landed on the branch just after the squash-merge, so it is NOT in main via #506 — it ships in follow-up #508 (cherry-picked, plus an ops 2.10.1 bump so it propagates to boxes already refreshed at 2.10.0).
